bug 1387236: extract and load study_state too

groovecoder opened the pull request:
I want to measure the % of users who submit 1+ "page-problem" pings. But we only have data for the 'page-problem|page-works|breakage|notes|disable' pings. So, we don't know how many users are running the study who have never sent a page problem/works report.

This PR adds `study_state` extraction and load into the table, so I can get the total # of clients running the study add-on - including the ones who have not sent any breakage reports at all.

groovecoder commented:
Updated description of PR.

groovecoder commented:
Tests failed here, but probably/maybe for a good reason? What happens to this job when the `study_state` field isn't included? (Which happens on the in-study pings vs. the shield-state pings)

groovecoder commented:
@acmiyaguchi or @sunahsuh - can I get an r? here? In case @harterrt isn't able to review this week?

acmiyaguchi commented on mozetl/shield/privacy_prefs.py:
```diff
@@ -23,6 +23,7 @@
     ("breakage", "payload/breakage", None, StringType()),
     ("notes", "payload/notes", None, StringType()),
     ("study", "payload/study", None, StringType()),
+    ("study_state", "payload/study_state", None, StringType()),
```
You should add `study_state` to the failing test inputs. From the travis logs:

```
>       assert row_to_dict(actual) == create_row(result_payload)
E       AssertionError: assert {'branch': 't...problem', ...} == {'branch': 'th...problem', ...}
E         Omitting 7 identical items, use -vv to show
E         Left contains more items:
E         {'study_state': None}
E         Use -v to get the full diff
```

If you omit the field, it still gets filled in as a null value because it's included with the table schema.

acmiyaguchi commented:
You should also formulate your question and assumptions about the inclusion of `study_state` as a test.

groovecoder commented:
Updated with more + better tests. I also introduce 2 separate transforms in the single ETL job: 1 to transform "state" pings and 1 to transform "event" pings. I believe I've written this correctly, but my tests for the new "state" pings are failing and I don't understand why. 😢 

### State Pings
These are sent by the underlying shield library. They use `study_name` instead of `study` and include `"study_state": "ineligible|install|running|user-ended-study"` values. E.g.,
```
{
  "study_name": "@shield-study-privacy",
  "branch": "firstPartyIsolationOpenerAccess",
  "study_state": "running",
  "study_version": "0.0.4",
  "about": {
    "_src": "shield",
    "_v": 2
  }
}
```

### Event pings
These are sent by *MY* study code. They use `study` and include `originDomain`, `event`, `breakage`, and `notes` fields & values. E.g.,
```
{
  "study": "@shield-study-privacy",
  "branch": "firstPartyIsolationOpenerAccess",
  "originDomain": "addons.mozilla.org",
  "event": "page-works",
  "breakage": "",
  "notes": "",
  "study_version": "0.0.4",
  "about": {
    "_src": "addon",
    "_v": 2
  }
}
```

acmiyaguchi commented on tests/test_shield_privacy_prefs.py:
```diff
@@ -12,16 +13,111 @@ def create_ping_rdd(sc, payload):
 
 
+def create_state_row(overrides):
+    keys = ["client_id", "branch", "study", "study_state"]
```
This should be `study_name` instead of `study`, as per your comment about state pings.

groovecoder replied:
When I did that, I got ...
```
>       assert row_to_dict(actual) == create_state_row(result_payload)
E       AssertionError: assert {'branch': 'f... 'ineligible'} == {'branch': 'fi... 'ineligible'}
E         Omitting 3 identical items, use -vv to show
E         Left contains more items:
E         {'study': u'@shield-study-privacy'}
E         Right contains more items:
E         {'study_name': '@shield-study-privacy'}
E         Use -v to get the full diff
```

That's why I'm confused. 😢

groovecoder replied:
But I think this might be the right behavior, and I need to change the test to match? I.e., I **WANT** the transform to turn the state ping payload `study_name` into `study` in parquet table so it's consistent with the event pings in parquet.

acmiyaguchi commented on tests/test_shield_privacy_prefs.py:
```diff
@@ -12,16 +13,111 @@ def create_ping_rdd(sc, payload):
+    return {key: overrides.get(key, None) for key in keys}
+
+
+def test_study_running(row_to_dict, spark_context):
```
There doesn't seem to be a code path that checks whether the `study_state` is within a given whitelist of options. These tests seem pretty redundant.

acmiyaguchi commented on mozetl/shield/utils.py:
```diff
@@ -17,7 +17,7 @@ def etl_job(sc, sqlContext, submission_date=None, save=True):
         transformed_pings = transform_func(sqlContext, pings)
 
-        if save:
+        if type(save) == bool and save is True:
```
What does does add?

groovecoder replied:
I had started to expand this functionality to accept save value of `True`, `overwrite`, or `____`? - i.e., whatever `write.mode()` does **NOT** overwrite existing records.

I'm worried that calling an `etl_job` twice from `privacy_prefs.py` with the same `s3_path` means the 2nd job will overwrite the first job's records?

acmiyaguchi replied:
From what I understand, you don't want to have two separate datasets to process this dataset. If so, you should be taking the union of the two sets before writing them to disk.

On another note, the option you probably want is `mode="error"`. https://spark.apache.org/docs/latest/sql-programming-guide.html#save-modes

groovecoder commented:
Okay, this needs a new review because I changed it pretty substantially. Is this the right way to combine `transformed_pings` together before writing them to parquet?
```
transformed_pings = transformed_event_pings + transformed_state_pings
```
?

acmiyaguchi commented:
Not quite, you should probably use [`union` ie `tranformed_event_pings.union(transformed_state_pings)`](http://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.DataFrame.union). You also have to reconcile the two dataframes because they now have very different schemas. I would definitely write a test that combines the result of those two datasets and runs a very simple sample query over them (similar to how you would use it in practice).

groovecoder commented:
Ready for next (last?) review ...

1. Updated to use `.union`
2. Consolidated `COMMON_COLUMN_CONFIGS` and used tiny `STUDY_STATE_DATAFRAME_COLUMN_CONFIGS|STUDY_EVENT_DATAFRAME_COLUMN_CONFIGS` to handle the different `study` paths
3. Added `test_combine_state_and_event` to test the union.

acmiyaguchi commented on tests/test_shield_privacy_prefs.py:
```diff
@@ -14,14 +15,105 @@ def create_ping_rdd(sc, payload):
     return {key: overrides.get(key, None) for key in keys}
 
 
+def test_study_running(row_to_dict, spark_context):
```
I still think these 4 tests are redundant. The only field that changes is `study_state`, but there is nothing that checks the validity of `study_state`. A better test would be an invalid or null `study_state`, plus a single valid case.

acmiyaguchi commented on mozetl/shield/privacy_prefs.py:
```diff
@@ -5,40 +5,85 @@
+    if submission_date is None:
+        submission_date = (date.today() - timedelta(1)).strftime("%Y%m%d")
+
+    pings = Dataset.from_source(
```
Nit: I think that using a new scope in parenthesis easier to read for long method chains.

```
(
    Dataset
    .from_source("telemetry")
    .where(
        docType="shield-study",
        submissionDate=submission_date,
        appName="Firefox",
    )
    .records(sc)
)
```

groovecoder commented:
Updated to remove redundant tests.
